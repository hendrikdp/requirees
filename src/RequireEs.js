import {getDefinitionArguments, getRequireArguments} from "./helpers/arguments.js";
import currentTagLoad from './helpers/currentTagLoad.js';
import Registry from './classes/Registry.js';
import Loaders from './classes/Loaders.js';
import {root} from "./require-global";
import {transformRJSPaths} from './helpers/requireJsTransformers.js';

export default class{

    constructor(options){
        this.registry = new Registry(options, this);
        this.options = options || {};
        this.amd = {};
        this.loaders = new Loaders(this);
    }

    define(){
        const {name, dependencies, factory} = getDefinitionArguments(arguments);
        //Carefull with jQuery: In general, explicitly naming modules in the define() call are discouraged, but jQuery has some special constraints.
        const isNamed = typeof name === 'string' && !(name==='jquery' && typeof factory === 'undefined');
        if(isNamed){
            const registryElement = {};
            registryElement[name] = {dependencies, factory};
            this.register(registryElement);
        }else{
            //if this is an anonymous define, confirm the currently loading script that loading is done...
            currentTagLoad.confirmDefine({dependencies, factory});
        }
    }

    get(){
        //todo: build in fail callback logic
        const {dependencies, callback, loadSinglePackage, options} = getRequireArguments(arguments);
        const targetInstances = dependencies.map(target => {
            let results = this.findOne(target);
            if(typeof results.match === 'undefined'){
                this.register(target, options);
                results = this.findOne(target);
            }
            return this.loaders.load(results, options);
        });
        const allScriptsLoaded = Promise.all(targetInstances);
        if(typeof callback === 'function'){
            allScriptsLoaded.then(instances => callback.apply(root, instances));
        }
        return loadSinglePackage ? targetInstances[0] : allScriptsLoaded;
    }

    getPromise(){
        return new Promise(resolve => resolve(this.get.apply(this, arguments)));
    }

    register(){
        return this.registry.add.apply(this.registry, arguments);
    }

    find(){
        return this.registry.find.apply(this.registry, arguments);
    }

    findOne(){
        return this.registry.findOne.apply(this.registry, arguments);
    }

    config(options = {}){
        if(typeof options.paths !== 'undefined') transformRJSPaths(options.paths).forEach(this.register.bind(this));
        if(typeof options.allowRedefine !== 'undefined') this.options.allowRedefine = options.allowRedefine;
    }

    specified(packageName){
        if(typeof packageName === 'string'){
            const result = this.findOne(packageName);
            return result.match && result.match.isSpecified(result.attrs) || false;
        }
        return false;
    }

    //returns the instance as a function
    asFunction(usePromises){
        const requirees = usePromises ? this.getPromise.bind(this) : this.get.bind(this);
        requirees.register = this.register.bind(this);
        requirees.find = this.find.bind(this);
        requirees.findOne = this.findOne.bind(this);
        requirees.loaders = this.loaders;
        requirees.define = this.define.bind(this);
        requirees.config = this.config.bind(this);
        requirees.specified = this.specified.bind(this);
        return requirees;
    }

}