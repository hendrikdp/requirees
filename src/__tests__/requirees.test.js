import RequireEs from '../RequireEs.js';

describe('Registry handling - redefine existing packages', ()=>{

    let requirees;
    let factory1;
    let factory2;

    beforeEach(()=>{
        requirees = new RequireEs();
        factory1 = ()=>{return {instance: 'factory1'}};
        factory2 = ()=>{return {instance: 'factory2'}};
    });


    test('Do not allow re-defines on packages which are already specified', async () => {
        requirees.define('pckg1', [], factory1);
        await requirees.get('pckg1');
        requirees.define('pckg1', [], factory2);
        const result = await requirees.get('pckg1');
        expect(result.instance).toBe('factory1');
    });

    test('Allow redefines if the factory has not ran yet', async () => {
        requirees.define('pckg1', [], factory1);
        requirees.define('pckg1', [], factory2);
        const result = await requirees.get('pckg1');
        expect(result.instance).toBe('factory2');
    });

    test('Allow re-defines on packages which are already specified, when specifically specified', async () => {
        requirees.define('pckg1', [], factory1);
        await requirees.get('pckg1');
        requirees.config({allowRedefine: true});
        requirees.define('pckg1', [], factory2);
        const result = await requirees.get('pckg1');
        expect(result.instance).toBe('factory2');
    });

});