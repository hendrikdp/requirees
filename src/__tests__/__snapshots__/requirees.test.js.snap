// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test RequireJs registrations + registration while getting Automatic registration while getting package 1`] = `
Object {
  "cdnjscloudflarecomajaxlibsreactumdreactproductionmin": Array [
    Object {
      "filetypes": Object {
        "js": Object {
          "dfr": Promise {},
          "urls": Array [
            "https://cdnjs.cloudflare.com/ajax/libs/react/16.14.0/umd/react.production.min.js",
          ],
        },
      },
      "major": 16,
      "minor": 14,
      "patch": 0,
      "str": "16.14.0",
    },
  ],
}
`;

exports[`Test shim functionality check if dependencies an be added to different filetypes 1`] = `
Object {
  "dependenciesPreLoad": Array [
    "dep1",
  ],
  "postFactory": [Function],
  "urls": Array [],
}
`;

exports[`Test shim functionality check if shim exports can be a function 1`] = `
Object {
  "postFactory": [Function],
  "urls": Array [],
}
`;

exports[`Test shim functionality check if shims are registered on unknown packages 1`] = `
Object {
  "js": Object {
    "dependenciesPreLoad": Array [
      "dep1",
      "dep2",
    ],
    "postFactory": [Function],
    "urls": Array [],
  },
}
`;
